Improve development workflow in docker

michaelarnauts opened the pull request:
* Merged some pip3 installs to one step in the `Dockerfile`.
* Added comments to the scripts in `script/`.
* Added `set -e` to all the scripts, so they don't continue when an error occurs.
* Correctly check out `python3` branch of openzwave in `build_python_openzwave`.
* Renamed `virtualization/Docker/Dockerfile.test` to `virtualization/Docker/Dockerfile.dev` since this image can be used for all development.
* Changed `virtualization/Docker/Dockerfile.dev` to match the production ´Dockerfile`. The old one was still running tests on python 3.4.
* Changed `script/bootstrap_frontend` to add `--allow-root` to allow installing bower packages as root. (See https://github.com/bower/bower/issues/1752)
* Changed LANG env in `tox.ini` to `LANG=C.UTF-8` since that is what seems to be available on most systems. (see https://github.com/docker-library/python/issues/13)

michaelarnauts committed:
Allow bower install of frontend components as root. Needed for frontend development in docker since everything runs as root in the docker image.

balloob commented:
Can you include a comment to the script why the flag is there

michaelarnauts commented:
Actually, I will document all scripts a bit, and make sure they handle errors fine. Currently, alsmost none of them do a `set -e`, so they all happily continue when an error occurs. This made it a lot harder to spot issues in my dev environment.

janLo commented:
Isn't it better, to just switch to another user before installing it in a docker container?

michaelarnauts commented:
Not sure if it's better. There is no security risk afaik, it's just to protect the user to install the components as root since you may encounter permission issues afterwards whenever you try to change them as a user.

michaelarnauts committed:
Improve development workflow in docker

michaelarnauts committed:
Use LANG=C.UTF-8 in tox. Fixes installation of libraries with UTF-8 in it's readme.

michaelarnauts commented:
I changed `tox.ini` to specify `LANG=C.UTF-8`. The `python:3.5` docker images doesn't contain `en_US.UTF-8` locale, and generating it only increases the docker image size while that is not necessary.

kellerza commented on virtualization/Docker/Dockerfile.dev:
```diff
@@ -0,0 +1,55 @@
+# Based on the production Dockerfile, but with development additions.
+# Keep this file as close as possible to the production Dockerfile, so the environments match.
+
+FROM python:3.5
```
Since our minimum required version is 3.4.2 (Raspberry PI), its probably better to keep developing there?

michaelarnauts replied:
I only changed the development Docker to match the production version. That was already changed to 3.5.

balloob replied:
Would it be possible to do `FROM homeassistant/home-assistant:dev` ?

michaelarnauts replied:
I would love to do something like that, but that would add a requirement on the Dockerfile on dockerhub. This would make it impossible to change anything in the Dockerfile on your local development branch, since the image from dockerhub would be fetched every time. This doesn't seem like a good idea.

A better approach would be to depend on the main Dockerfile, but docker doesn't want to implement this (see https://github.com/docker/docker/issues/735).

michaelarnauts commented on Dockerfile:
```diff
@@ -6,7 +6,7 @@ VOLUME /config
 WORKDIR /usr/src/app
 
-RUN pip3 install --no-cache-dir colorlog cython
+RUN pip3 install --no-cache-dir colorlog cython mysqlclient psycopg2 uvloop
```
I don't know about these dependencies, so maybe they can just move to the requirements.txt file?

kellerza replied:
All three requirements are optional (for DB and uvloop is a specific async library used on some platforms, like Docker). 

Lets just keep their install after the requirements.txt

The requirements I'm talking about are the requirements of `uvloop` for example, it requires `aiohttp` (any version), and we pin `aiohttp`

michaelarnauts committed:
Install mysqlclient psycopg2 uvloop after requirements_all.txt again, but with a --no-cache-dir this time. Allows bootstrap_frontend to be executed in a different path like the other scripts.

michaelarnauts commented:
@kellerza As requested, I've moved the installation of `mysqlclient psycopg2 uvloop` back to after the `requirements_all.txt`, but I've added a `--no-cache-dir` to keep the docker layers clean.

kellerza commented:
Looks good @michaelarnauts ! 🐬 

I only run my production on Docker, so will leave merging for @balloob 

Can we create a `hass` script in the Dockerfile? I currently access bash in Docker using `docker exec -ti home-assistant bash` and would be good to have `hass` available as a shortcut then.

The main reason I don't use it for `dev` is after each Docker update you will have to point git at your repo etc. So additional docs on your workflow would be nice.

balloob commented on virtualization/Docker/Dockerfile.dev:
```diff
@@ -0,0 +1,56 @@
+# BEGIN: Development additions
+
+# Install nodejs
+RUN curl -sL https://deb.nodesource.com/setup_7.x | sudo -E bash - && \
```
Instead of installing node 7, we should use nvm to install the node that is specified in `.nvmrc`

michaelarnauts replied:
Only that the `.nvmrc` is in the polymer project, and that project might not be checked out when the image is build. Can this file be moved to the main project?

balloob replied:
No, it's specific to the polymer project.

I actually don't see any checking out of the polymer project so node is not necessary?

michaelarnauts replied:
That's true, but I want to have node installed in the development docker image, so I didn't need to install node every time my image was rebuild (what happens on every execution of `script/dev_docker` since the code changes).

The dev image also needs to work without the polymer project checked out, so unless you check out the polymer project, the `.nvmrc` file is not there, right?

This all in a quest to make development on hass in docker just work, without worrying on dependencies and build tools.

michaelarnauts commented:
@kellerza You could simply create an [alias](http://www.linfo.org/alias.html) for that, so you don't need to type the full docker exec command :)

Also, you should add an additional origin with a link to your repo. I did it like this:
```shell
$ git remote add michaelarnauts https://github.com/michaelarnauts/home-assistant.git
```

Then, when I need to push, i just have to specify the origin:
```shell
$ git push michaelarnauts <branch>
```

A `git pull` will still update from the main home-assistant repo. Makes developing in docker easy, and keeps my system clean of python dependencies.
